File: love2d/src/agent/session.lua
-- One conversation with the model, and the loop that lets it act.
--
-- `ask()` sends the transcript, streams the prose back, and — while the model
-- keeps asking for tools — runs them on the bench and sends the results, up to
-- `MAX_ROUNDS` turns. That is the whole "coding agent": read the file, edit or
-- write it, run it, read what the compiler said, fix it, run again, then
-- explain. Every step of it happens on this machine (docs/agent.md §3).
--
-- The session is the client's memory of the room: what was said, which tools
-- ran, what they answered. The screen posts the human-readable messages to the
-- server for keeps (`playground.chat.post`); this holds the shape the model
-- needs, tool ids and all, for the length of the visit.
--
-- ## A coroutine, not a promise
--
-- The browser writes this with `await`. Here the whole ask runs inside one
-- coroutine that yields a frame whenever it is waiting — on the network, on
-- the typist, on a run — and `Session:update` resumes it once a frame. The
-- code either side of a wait is therefore the same straight line in both
-- clients. (LuaJIT can yield across `pcall`, which is what lets `tools.lua`
-- guard a tool that waits.)

local Prefs = require("src.agent.prefs")
local Providers = require("src.agent.providers")
local Tools = require("src.agent.tools")
local json = require("src.json")

local M = {}

--- How many times in one ask the model may come back for another tool.
M.MAX_ROUNDS = 10
--- Transcript turns kept for the model; older ones fall off the front.
M.KEEP_TURNS = 24

--- The names the game uses for itself, for the model.
local LANG_NAME = {
  rust = "Rust",
  go = "Go",
  cpp = "C++ (C++20)",
  python = "Python 3",
  pytorch = "Python 3 + PyTorch",
  typescript = "TypeScript (tsc --strict, run on Node)",
}

--- The most of one stream or one wrong answer that goes into the prompt.
local CLIP = 2000

local function clip(text)
  if #text <= CLIP then return text end
  return text:sub(1, CLIP) .. ("\n… (%d more characters)"):format(#text - CLIP)
end

--- JSON's spelling of a string, for the sample cases: a newline has to be
--- visible in the prompt as `\n` or a one-line expectation reads as two.
local function quoted(text)
  return json.encode(tostring(text or ""))
end

--- The exercise, for the screens that set one (`bench.task`).
---
--- Everything in here is something the person is already looking at: the
--- brief they are reading, the sample cases under it, the report the last RUN
--- printed, and — only on a quest they have already cleared, because that is
--- the only time the server sends it — the reference answer. The hidden cases
--- are a number, the way they are on the screen.
local function task_lines(task, out)
  out[#out + 1] = ""
  out[#out + 1] =
    "The person is not writing whatever they like: this screen is a graded exercise, and this is it."
  out[#out + 1] = ""
  out[#out + 1] = ("Title: %s"):format(task.title or "")
  out[#out + 1] = ("Brief: %s"):format(task.brief or "")
  if task.story and task.story:match("%S") then
    out[#out + 1] = ("Story: %s"):format(task.story)
  end
  local tests = task.tests or {}
  if #tests > 0 then
    out[#out + 1] = ""
    out[#out + 1] = "The sample cases (stdin → expected stdout):"
    for _, c in ipairs(tests) do
      out[#out + 1] = ("- %s → %s"):format(quoted(c.stdin), quoted(c.expect))
    end
  end
  if (task.hidden_count or 0) > 0 then
    out[#out + 1] = ("There are also %d hidden case(s) nobody can see, including you. A program that only answers the samples will fail them.")
      :format(task.hidden_count)
  end
  local run = task.last_run
  if run then
    out[#out + 1] = ""
    out[#out + 1] = ("The last RUN: %s, %d/%d of the visible cases passed.")
      :format(run.verdict or "?", run.passed or 0, run.total or 0)
    if run.stderr and run.stderr:match("%S") then
      out[#out + 1] = "What the compiler or the program said:"
      out[#out + 1] = "```"
      out[#out + 1] = clip(run.stderr)
      out[#out + 1] = "```"
    end
    for _, c in ipairs(run.failing or {}) do
      out[#out + 1] = ("Failed on %s: expected %s, got %s.")
        :format(quoted(c.stdin), quoted(c.expect), quoted(clip(c.got or "")))
    end
  end
  if task.cleared then
    out[#out + 1] = ""
    out[#out + 1] =
      "The person has already cleared this one; they are here to practise or to understand it."
  end
  if task.solution and task.solution:match("%S") then
    out[#out + 1] = ""
    out[#out + 1] =
      "The reference answer (they have cleared this quest, so they can already read it — use it to explain, not to replace their own):"
    out[#out + 1] = "```"
    out[#out + 1] = task.solution
    out[#out + 1] = "```"
  end
  out[#out + 1] = ""
  out[#out + 1] =
    "On a graded screen: answer the question they asked. Explain, name the line, say what the compiler is complaining about. They learn nothing from a program that appeared while they were reading, so write the whole answer into the editor only when they ask you for it outright — and then say what it does."
end

--- The coder's standing orders. Stable text first, the exercise, the file last.
---
--- Word-for-word the browser's (`frontend/src/ai/session.ts`). The prompt is
--- what makes the character; two clients with different prompts would be two
--- different characters wearing the same sprite.
function M.system_prompt(bench, can_run)
  local lang = LANG_NAME[bench.lang] or bench.lang
  local task = bench.task and bench.task() or nil
  local lines = {
    ("You are the Rust coder — a small pixel-art character on a flying keyboard who lives on the code screen of Causewaybay Hacker, a 16-bit coding game set in Hong Kong. The person is learning %s. You are their pair: a friendly senior engineer who explains briefly and lets the code speak."):format(lang),
    "",
    "The whole project is ONE source file, the one in the editor. There are no other files, no build system to configure, no dependencies beyond the standard library (Rust: std only, no crates; Go: standard library; C++: the standard library, compiled with -std=c++20; Python 3: the standard library; TypeScript: tsc in strict mode with no @types/node — only process, console, the timers and fs.readFileSync are declared, so stdin is read with `const input: string = require(\"fs\").readFileSync(0, \"utf8\");` and output goes through console.log).",
    "",
    "How to work:",
    "- For a small change, use edit_code with the exact span. For a new program or a rewrite, use write_code. Both are typed into the editor character by character while the person watches, so write only what is needed and no filler comments.",
    can_run
        and "- After changing code, call run_code and read the outcome. If it did not compile or crashed, read the compiler's message, fix the code, and run again — up to a few times — before you answer."
      or "- This screen cannot run code, so read carefully and reason about it instead.",
    "- Do not paste code into your prose when you could put it in the editor with a tool. Your prose is a short chat message: what you did and why, one to four sentences, plain text, no markdown headings.",
    "- If asked to review, be concrete: name the line and the habit, and offer the fix. Do not rewrite a working program nobody asked you to rewrite.",
    "- Keep the person's style, names and formatting unless asked. Keep the program's existing behaviour unless asked.",
    "- Never invent an API. If unsure, prefer the plain standard-library way.",
    "- A picture is not a program. When the person asks for a picture, image, drawing or photo, call make_image with a prompt; do not write code that prints one, and do not describe it instead.",
  }
  if task then task_lines(task, lines) end
  local tail = {
    "",
    ("The file is %s. Its current text, with line numbers (do not include the numbers in edits):"):format(bench.file),
    "```",
    Tools.numbered(bench.read()),
    "```",
  }
  for _, line in ipairs(tail) do lines[#lines + 1] = line end
  return table.concat(lines, "\n")
end

--- The last-round nudge: no tools left, so say where you got to.
M.FINAL_NUDGE =
  "You have used every tool round you have. Do not call tools; say in two or three sentences what you did and what, if anything, is left."

local Session = {}
Session.__index = Session

--- `listener` wants `text(delta)`, `tool(name, input)`,
--- `tool_done(name, text, error)` and `mood(m)`, where a mood is one of
--- "idle", "thinking", "typing", "running".
function M.new(bench, listener, wallet, lib)
  return setmetatable({
    bench = bench,
    listener = listener,
    wallet = wallet,
    lib = lib,
    messages = {},
    co = nil,
    stopping = false,
    last = "",
  }, Session)
end

function Session:busy()
  return self.co ~= nil
end

--- Forget the conversation. The server's copy, if any, is the screen's business.
function Session:clear()
  self:stop()
  self.messages = {}
end

--- Ask it to stop. The coroutine notices at its next wait: mid-stream the
--- request is cancelled, mid-typing the typist is stopped by the screen.
function Session:stop()
  if self.co then self.stopping = true end
end

local function push_text(messages, role, text)
  messages[#messages + 1] = { role = role, content = { { type = "text", text = text } } }
end

--- Keep the transcript to a size a model can hold, without ever cutting
--- between a tool call and its result: the front is dropped in whole
--- user-started exchanges.
function Session:trim()
  while #self.messages > M.KEEP_TURNS do
    table.remove(self.messages, 1)
    while #self.messages > 0 do
      local first = self.messages[1]
      local starts = false
      if first.role == "user" then
        for _, part in ipairs(first.content) do
          if part.type == "text" then starts = true end
        end
      end
      if starts then break end
      table.remove(self.messages, 1)
    end
  end
end

--- One ask, however many rounds it takes.
---
--- Returns immediately: the work runs in a coroutine that `update` pumps.
--- `done(text, err)` is called at the end with the prose of the last turn, or
--- nil and a message when the provider refused or the network failed.
function Session:ask(provider, text, done)
  if self:busy() then
    if done then done(nil, "busy") end
    return
  end
  local key = Prefs.key(provider)
  local model = Prefs.model(provider)
  if key == "" and Prefs.needs_key(provider) then
    if done then done(nil, "no api key") end
    return
  end
  self.stopping = false
  self.last = ""
  push_text(self.messages, "user", text)
  self:trim()
  local tools = Tools.tools_for(self.bench)
  local session = self

  self.co = coroutine.create(function()
    for round = 0, M.MAX_ROUNDS do
      session.listener.mood("thinking")
      -- The last round goes out with no tools at all, so a model that would
      -- keep going is made to stop and say where it got to rather than being
      -- cut off mid-loop with nothing said.
      local final = round == M.MAX_ROUNDS
      local messages = session.messages
      if final then
        messages = {}
        for i, m in ipairs(session.messages) do messages[i] = m end
        push_text(messages, "user", M.FINAL_NUDGE)
      end
      local turn, err = Providers.chat({
        wallet = session.wallet,
        lib = session.lib,
        provider = provider,
        key = key,
        model = model,
        system = M.system_prompt(session.bench, session.bench.run ~= nil),
        messages = messages,
        tools = final and {} or tools,
        on_text = function(delta) session.listener.text(delta) end,
        cancelled = function() return session.stopping end,
      })
      if not turn then
        if err == "stopped" then return session.last end
        error(err or "the provider said nothing", 0)
      end

      local parts = {}
      if turn.text ~= "" then parts[#parts + 1] = { type = "text", text = turn.text } end
      for _, use in ipairs(turn.tool_uses) do
        parts[#parts + 1] =
          { type = "tool_use", id = use.id, name = use.name, input = use.input }
      end
      if #parts > 0 then
        session.messages[#session.messages + 1] = { role = "assistant", content = parts }
      end
      session.last = turn.text
      if turn.stop ~= "tool" or #turn.tool_uses == 0 then return session.last end

      local results = {}
      for _, use in ipairs(turn.tool_uses) do
        if session.stopping then break end
        local mood = "thinking"
        if use.name == "run_code" then
          mood = "running"
        elseif use.name:match("code$") then
          mood = "typing"
        end
        session.listener.mood(mood)
        session.listener.tool(use.name, use.input)
        local out, failed
        if use.input.__invalid_json then
          out, failed = "The tool call's JSON did not parse; send it again.", true
        else
          out, failed = Tools.run_tool(session.bench, use.name, use.input)
        end
        session.listener.tool_done(use.name, out, failed)
        results[#results + 1] = {
          type = "tool_result",
          tool_use_id = use.id,
          name = use.name,
          content = out,
          is_error = failed,
        }
      end
      -- STOP during a tool: every tool_use in the assistant turn still has
      -- to be answered, or the next ask goes out with a call and no result
      -- and the provider refuses the whole transcript until the room is
      -- cleared. The tools that ran keep their result; the rest are told so.
      if session.stopping then
        for i = #results + 1, #turn.tool_uses do
          local use = turn.tool_uses[i]
          results[#results + 1] = {
            type = "tool_result",
            tool_use_id = use.id,
            name = use.name,
            content = "Stopped by the player before this tool ran.",
            is_error = true,
          }
        end
        session.messages[#session.messages + 1] = { role = "user", content = results }
        return session.last
      end
      session.messages[#session.messages + 1] = { role = "user", content = results }
    end
    return session.last
  end)

  self.done = done
  -- One resume now, so a failure that needs no network is reported this frame.
  self:update()
end

--- Give the ask a frame. Does nothing when there is none.
function Session:update()
  if not self.co then return end
  local ok, value = coroutine.resume(self.co)
  if coroutine.status(self.co) == "dead" then
    local done = self.done
    self.co = nil
    self.done = nil
    self.stopping = false
    self.listener.mood("idle")
    if done then
      if ok then
        done(value or "", nil)
      else
        done(nil, tostring(value))
      end
    end
  elseif not ok then
    -- A coroutine that is neither dead nor fine cannot happen; treated as a
    -- failure rather than left running.
    self.co = nil
    self.listener.mood("idle")
    if self.done then self.done(nil, tostring(value)) end
    self.done = nil
  end
end

M.Session = Session

return M

File: frontend/src/ai/session.ts
/**
 * One conversation with the model, and the loop that lets it act.
 *
 * `ask()` sends the transcript, streams the prose back, and — while the model
 * keeps asking for tools — runs them on the bench and sends the results, up
 * to `MAX_ROUNDS` turns. That is the whole "coding agent": read the file,
 * edit or write it, run it, read what the compiler said, fix it, run again,
 * then explain. Every step of it is in the browser (docs/agent.md §3).
 *
 * The session is the client's memory of the room: what was said, which tools
 * ran, what they answered. The screen posts the human-readable messages to
 * the server for keeps (`playground.chat.post`); this holds the shape the
 * model needs, tool ids and all, for the length of the visit.
 */
import { needsKey, readKey, readModel, type Provider } from "./prefs";
import { chat, type Msg, type Part, type Turn } from "./providers";
import { numbered, runTool, toolsFor, type Bench, type TaskBrief } from "./tools";

/** How many times in one ask the model may come back for another tool. */
export const MAX_ROUNDS = 10;
/** Transcript turns kept for the model; older ones fall off the front. */
const KEEP_TURNS = 24;

export type Mood = "idle" | "thinking" | "typing" | "running";

export interface Listener {
  /** A piece of the reply's prose. */
  text(delta: string): void;
  /** The model is about to run this tool. */
  tool(name: string, input: Record<string, unknown>): void;
  /** And this is what it answered. */
  toolDone(name: string, text: string, error: boolean): void;
  mood(m: Mood): void;
}

/** The names the game uses for itself, for the model. */
const LANG_NAME: Record<string, string> = {
  rust: "Rust",
  go: "Go",
  cpp: "C++ (C++20)",
  python: "Python 3",
};

/** The most of one stream or one wrong answer that goes into the prompt. */
const CLIP = 2000;

const clip = (text: string): string =>
  text.length > CLIP ? `${text.slice(0, CLIP)}\n… (${text.length - CLIP} more characters)` : text;

/**
 * The exercise, for the screens that set one (`Bench.task`).
 *
 * Everything in here is something the person is already looking at: the
 * brief they are reading, the sample cases under it, the report the last RUN
 * printed, and — only on a quest they have already cleared, because that is
 * the only time the server sends it — the reference answer. The hidden cases
 * are a number, the way they are on the screen; there is nothing here for an
 * unsolved quest's answer to come through, because the client does not have
 * it either.
 */
function taskLines(task: TaskBrief): string[] {
  const out = [
    "",
    "The person is not writing whatever they like: this screen is a graded exercise, and this is it.",
    "",
    `Title: ${task.title}`,
    `Brief: ${task.brief}`,
  ];
  if (task.story?.trim()) out.push(`Story: ${task.story}`);
  if (task.tests.length > 0) {
    out.push("", "The sample cases (stdin → expected stdout):");
    for (const c of task.tests)
      out.push(`- ${JSON.stringify(c.stdin)} → ${JSON.stringify(c.expect)}`);
  }
  if (task.hiddenCount > 0) {
    out.push(
      `There are also ${task.hiddenCount} hidden case(s) nobody can see, including you. A program that only answers the samples will fail them.`,
    );
  }
  if (task.lastRun) {
    const r = task.lastRun;
    out.push("", `The last RUN: ${r.verdict}, ${r.passed}/${r.total} of the visible cases passed.`);
    if (r.stderr.trim())
      out.push("What the compiler or the program said:", "```", clip(r.stderr), "```");
    for (const c of r.failing) {
      out.push(
        `Failed on ${JSON.stringify(c.stdin)}: expected ${JSON.stringify(c.expect)}, got ${JSON.stringify(clip(c.got))}.`,
      );
    }
  }
  if (task.cleared)
    out.push(
      "",
      "The person has already cleared this one; they are here to practise or to understand it.",
    );
  if (task.solution?.trim()) {
    out.push(
      "",
      "The reference answer (they have cleared this quest, so they can already read it — use it to explain, not to replace their own):",
      "```",
      task.solution,
      "```",
    );
  }
  out.push(
    "",
    "On a graded screen: answer the question they asked. Explain, name the line, say what the compiler is complaining about. They learn nothing from a program that appeared while they were reading, so write the whole answer into the editor only when they ask you for it outright — and then say what it does.",
  );
  return out;
}

/** The agent's standing orders. Stable text first, the exercise, the file last. */
export function systemPrompt(bench: Bench, canRun: boolean): string {
  const lang = LANG_NAME[bench.lang] ?? bench.lang;
  const task = bench.task?.() ?? null;
  return [
    `You are the Rust coder — a small pixel-art character on a flying keyboard who lives on the code screen of Causewaybay Hacker, a 16-bit coding game set in Hong Kong. The person is learning ${lang}. You are their pair: a friendly senior engineer who explains briefly and lets the code speak.`,
    "",
    "The whole project is ONE source file, the one in the editor. There are no other files, no build system to configure, no dependencies beyond the standard library (Rust: std only, no crates; Go: standard library; C++: the standard library, compiled with -std=c++20; Python 3: the standard library).",
    "",
    "How to work:",
    "- For a small change, use edit_code with the exact span. For a new program or a rewrite, use write_code. Both are typed into the editor character by character while the person watches, so write only what is needed and no filler comments.",
    canRun
      ? "- After changing code, call run_code and read the outcome. If it did not compile or crashed, read the compiler's message, fix the code, and run again — up to a few times — before you answer."
      : "- This screen cannot run code, so read carefully and reason about it instead.",
    "- Do not paste code into your prose when you could put it in the editor with a tool. Your prose is a short chat message: what you did and why, one to four sentences, plain text, no markdown headings.",
    "- If asked to review, be concrete: name the line and the habit, and offer the fix. Do not rewrite a working program nobody asked you to rewrite.",
    "- Keep the person's style, names and formatting unless asked. Keep the program's existing behaviour unless asked.",
    "- Never invent an API. If unsure, prefer the plain standard-library way.",
    "- A picture is not a program. When the person asks for a picture, image, drawing or photo, call make_image with a prompt; do not write code that prints one, and do not describe it instead.",
    ...(task ? taskLines(task) : []),
    "",
    `The file is ${bench.file}. Its current text, with line numbers (do not include the numbers in edits):`,
    "```",
    numbered(bench.read()),
    "```",
  ].join("\n");
}

export class Session {
  /** The transcript as the model sees it. */
  messages: Msg[] = [];
  private abort: AbortController | null = null;
  /**
   * True from the top of `ask` to its `finally`. Not derived from `abort`:
   * STOP aborts the loop, but the loop is still awaiting whatever tool it
   * was in (a run on the server, a picture), and until that returns and
   * its `tool_result` is on the transcript, a new ask would push its text
   * into the middle of the exchange and every provider would refuse the
   * lot.
   */
  private running = false;
  get busy(): boolean {
    return this.running;
  }

  constructor(
    private readonly bench: Bench,
    private readonly listener: Listener,
  ) {}

  /** Forget the conversation. The server's copy, if any, is the screen's business. */
  clear(): void {
    this.stop();
    this.messages = [];
  }

  /** Abort the ask in flight. `busy` stays true until its loop has unwound. */
  stop(): void {
    this.abort?.abort();
  }

  /**
   * One ask, however many rounds it takes. Resolves to the prose of the last
   * turn. Throws on a provider error (a bad key, a network failure); a model
   * refusal or a truncated answer resolves with what there is.
   */
  async ask(provider: Provider, text: string): Promise<string> {
    if (this.busy) throw new Error("busy");
    const key = readKey(provider);
    const model = readModel(provider);
    if (!key && needsKey(provider)) throw new Error("no api key");
    const abort = new AbortController();
    this.abort = abort;
    this.messages.push({ role: "user", content: [{ type: "text", text }] });
    this.trim();
    // The transcript this ask is part of. CLEAR mid-tool swaps `this.messages`
    // for an empty one; the results still coming back belong to the old
    // exchange and must not land at the front of the new.
    const messages = this.messages;
    const tools = toolsFor(this.bench);
    let last = "";
    this.running = true;
    try {
      for (let round = 0; round <= MAX_ROUNDS; round++) {
        this.listener.mood("thinking");
        // The last round goes out with no tools at all, so a model that
        // would keep going is made to stop and say where it got to
        // (CausewaybayOffice's two-stage cap) rather than being cut off
        // mid-loop with nothing said.
        const final = round === MAX_ROUNDS;
        const turn: Turn = await chat({
          provider,
          key,
          model,
          system: systemPrompt(this.bench, this.bench.run !== null),
          messages: final
            ? [
                ...messages,
                {
                  role: "user",
                  content: [
                    {
                      type: "text",
                      text: "You have used every tool round you have. Do not call tools; say in two or three sentences what you did and what, if anything, is left.",
                    },
                  ],
                },
              ]
            : messages,
          tools: final ? [] : tools,
          signal: abort.signal,
          onText: (d) => this.listener.text(d),
        });
        const parts: Part[] = [];
        if (turn.text) parts.push({ type: "text", text: turn.text });
        for (const u of turn.toolUses) parts.push({ type: "tool_use", ...u });
        if (parts.length) messages.push({ role: "assistant", content: parts });
        last = turn.text;
        if (turn.toolUses.length === 0) break;
        // A turn cut short (the token cap, a refusal) can still carry a
        // tool call the model never got to finish. It does not run, but it
        // has to be answered all the same, or the transcript is invalid.
        if (turn.stop !== "tool") {
          messages.push({
            role: "user",
            content: turn.toolUses.map((u): Part => ({
              type: "tool_result",
              tool_use_id: u.id,
              name: u.name,
              content: "The reply was cut off before this tool could run.",
              is_error: true,
            })),
          });
          break;
        }
        const results: Part[] = [];
        for (const u of turn.toolUses) {
          if (abort.signal.aborted) break;
          const mood: Mood =
            u.name === "run_code" ? "running" : /code$/.test(u.name) ? "typing" : "thinking";
          this.listener.mood(mood);
          this.listener.tool(u.name, u.input);
          const r =
            "__invalid_json" in u.input
              ? { text: "The tool call's JSON did not parse; send it again.", error: true }
              : await runTool(this.bench, u.name, u.input);
          this.listener.toolDone(u.name, r.text, r.error);
          results.push({
            type: "tool_result",
            tool_use_id: u.id,
            name: u.name,
            content: r.text,
            is_error: r.error,
          });
        }
        // STOP during a tool: every `tool_use` in the assistant turn still
        // has to be answered, or the next ask goes out with a call and no
        // result and every provider refuses the whole transcript until the
        // room is cleared. The tools that ran keep their real result; the
        // ones that did not get told so.
        if (abort.signal.aborted) {
          for (const u of turn.toolUses.slice(results.length)) {
            results.push({
              type: "tool_result",
              tool_use_id: u.id,
              name: u.name,
              content: "Stopped by the player before this tool ran.",
              is_error: true,
            });
          }
          messages.push({ role: "user", content: results });
          break;
        }
        messages.push({ role: "user", content: results });
      }
    } finally {
      if (this.abort === abort) this.abort = null;
      this.running = false;
      this.listener.mood("idle");
    }
    return last;
  }

  /**
   * Keep the transcript to a size a model can hold, without ever cutting
   * between a tool call and its result: the front is dropped in whole
   * user-started exchanges.
   */
  private trim(): void {
    while (this.messages.length > KEEP_TURNS) {
      // Drop through to the next user message that carries text, which is
      // where the next exchange starts.
      this.messages.shift();
      while (
        this.messages.length > 0 &&
        !(
          this.messages[0].role === "user" &&
          this.messages[0].content.some((p) => p.type === "text")
        )
      ) {
        this.messages.shift();
      }
    }
  }
}
